File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章选数.py
#选数  lanqiaoOJ 753
from math import *
from itertools import *

def is_prime(n):
    if n == 1: return False
    m = int(sqrt(n)+1)
    for i in range(2,m):
        if n % i == 0:  return False
    return True

n,k = map(int,input().split())
s = list(map(int,input().split()))
cnt = 0
for e in combinations(s, k):            #所有组合
    num = sum(e)                  #求和
    if is_prime(num) == True:  cnt+=1
print(cnt)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章选数.cpp
//选数  lanqiaoOJ 753
#include<bits/stdc++.h>
using namespace std;
int n,m,b[25],a, cnt, c[25];
bool is_prime(int n){
     if(n <= 1)   return false;         //1不是素数
     for(long long i=2; i <= sqrt(n); i++)
         if(n % i == 0)  return false;  //能整除，不是素数
     return true;                       //读者思考，n=2时返回true吗？
}
void dfs(int dep,int e){
    if(dep==m) {
        if(is_prime(a)){
            cnt++;
            return ;
        }
        return ;
    }
    for(int i=e;i<=n;i++)//搜索
      if(c[i]==0)  {
         c[i]=1;
         a+=b[i];//相加
         dfs(dep+1,i+1);
         a-=b[i];//回溯
         c[i]=0;
     }
}
int main(){
    cin>>n>>m;
    for(int i=1;i<=n;i++)   cin>>b[i];
    dfs(0,1);
    cout<<cnt;
  return 0;
}
